pkg/asset/machines/master.go

```diff
@@ -114,9 +135,108 @@ func (m *Master) Generate(dependencies asset.Parents) error {
 	default:
 		return fmt.Errorf("invalid Platform")
 	}
+
+	machines := []interface{}{}
+	if m.Machines == nil {
+		for _, machine := range m.MachinesDeprecated {
+			machines = append(machines, machine)
+		}
+	} else {
+		for _, machine := range m.Machines {
+			machines = append(machines, machine)
+		}
+	}
+
+	for i, machine := range machines {
+		data, err := yaml.Marshal(machine)
+		if err != nil {
+			return errors.Wrapf(err, "marshal master %d", i)
+		}
+
+		m.FileList = append(m.FileList, &asset.File{
+			Filename: filepath.Join(directory, fmt.Sprintf(MasterMachineFileName, strconv.Itoa(i))),
+			Data:     data,
+		})
+	}
+
+	asset.SortFiles(m.FileList)
+
 	return nil
 }
 
+// Files returns the files generated by the asset.
+func (m *Master) Files() []*asset.File {
+	return m.FileList
+}
+
+// Load reads the asset files from disk.
+func (m *Master) Load(f asset.FileFetcher) (found bool, err error) {
+	file, err := f.FetchByName(filepath.Join(directory, MasterUserDataFileName))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	m.FileList = []*asset.File{file}
+
+	fileList, err := f.FetchByPattern(filepath.Join(directory, fmt.Sprintf(MasterMachineFileName, "*")))
+	if err != nil {
+		return true, err
+	}
+	m.FileList = append(m.FileList, fileList...)
+	for _, file := range fileList {
+		machine := &machineapi.Machine{}
+		err = yaml.Unmarshal(file.Data, &machine)
+		if err != nil {
+			return true, errors.Wrapf(err, "unmarshal %q", file.Filename)
+		}
+
+		rawProvider := machine.Spec.ProviderSpec.Value.Raw
+		providerType := &runtime.TypeMeta{}
+		err = yaml.Unmarshal(rawProvider, &providerType)
+		if err != nil {
+			return true, errors.Wrapf(err, "unmarshal %q provider type", file.Filename)
+		}
+
+		switch providerType.APIVersion {
+		case "awsproviderconfig.k8s.io/v1alpha1":
+			provider := &awsprovider.AWSMachineProviderConfig{}
+			err = yaml.Unmarshal(rawProvider, &provider)
+			if err != nil {
+				return true, errors.Wrapf(err, "unmarshal %s AWS provider", file.Filename)
+			}
+			machine.Spec.ProviderSpec.Value = &runtime.RawExtension{Object: provider}
+			m.Machines = append(m.Machines, *machine)
```

Comment:
If there are more than 10 machines, then the order of the machines in the slice will not be preserved when loaded back in. This will cause the asset store to think that the asset is dirty.

Reply:
> If there are more than 10 machines, then the order of the machines in the slice will not be preserved when loaded back in.

I sort [here][1] and [here][2] to guard against that.

[1]: https://github.com/wking/openshift-installer/commit/d5c01c0ef2d016b30ba750e97bc17db96f30f129#diff-66b4febb36af0238bb6c09e72c1c44ecR165
[2]: https://github.com/wking/openshift-installer/commit/d5c01c0ef2d016b30ba750e97bc17db96f30f129#diff-66b4febb36af0238bb6c09e72c1c44ecR241

Reply:
Those sort `FileList` but do nothing for `Machines` or `MachinesDeprecated`.

Let's say that there are 11 machines. The machines will be loaded into `Machines` in this order: 0, 1, 10, 2, 3, 4, 5, 6, 7, 8, 9. That will be different than the numerical order that they were in when the asset was generated.

Reply:
> Those sort FileList but do nothing for `Machines` or `MachinesDeprecated`.

Ah, right.  I've pushed 1c0796a -> 24aa42f addressing this.  Notes on the padded-index calculations in the new commit message.